sdks/java/core/src/main/java/org/apache/beam/sdk/util/ShardedKey.java

```diff
@@ -0,0 +1,117 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import com.google.auto.value.AutoValue;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.BooleanCoder;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.StructuredCoder;
+import org.apache.beam.sdk.util.common.ElementByteSizeObserver;
+
+/**
+ * A sharded key consisting of a user key and a shard id represented by bytes.
+ *
+ * <p>This is a more generic definition of {@link org.apache.beam.sdk.values.ShardedKey}.
+ */
+@AutoValue
+public abstract class ShardedKey<K> {
+
+  public static <K> ShardedKey<K> of(K key, @Nullable byte[] shardId) {
+    return new AutoValue_ShardedKey(shardId, key);
+  }
+
+  @SuppressWarnings("mutable")
+  @Nullable
+  public abstract byte[] getShardId();
+
+  public abstract K getKey();
+
+  public static class Coder<K> extends StructuredCoder<ShardedKey<K>> {
+
+    private final ByteArrayCoder shardCoder = ByteArrayCoder.of();
+    private final org.apache.beam.sdk.coders.Coder<K> keyCoder;
+
+    private Coder(org.apache.beam.sdk.coders.Coder<K> coder) {
+      keyCoder = coder;
+    }
+
+    public static <K> ShardedKey.Coder<K> of(org.apache.beam.sdk.coders.Coder<K> keyCoder) {
+      return new ShardedKey.Coder<K>(keyCoder);
+    }
+
+    public org.apache.beam.sdk.coders.Coder<K> getKeyCoder() {
+      return keyCoder;
+    }
+
+    @Override
+    public void encode(ShardedKey<K> shardedKey, OutputStream outStream) throws IOException {
+      // The encoding should follow the order:
+      //   null indicator
+      //   length of shard id
+      //   shard id
+      //   encoded user key
+      BooleanCoder.of().encode(shardedKey.getShardId() != null, outStream);
+      if (shardedKey.getShardId() != null) {
+        shardCoder.encode(shardedKey.getShardId(), outStream);
+      }
+      keyCoder.encode(shardedKey.getKey(), outStream);
+    }
+
+    @Override
+    public ShardedKey<K> decode(InputStream inStream) throws IOException {
+      if (!BooleanCoder.of().decode(inStream)) {
+        K key = keyCoder.decode(inStream);
+        return ShardedKey.of(key, null);
+      }
+      byte[] shardId = shardCoder.decode(inStream);
+      K key = keyCoder.decode(inStream);
+      return ShardedKey.of(key, shardId);
+    }
+
+    @Override
+    public List<? extends org.apache.beam.sdk.coders.Coder<?>> getCoderArguments() {
+      return Collections.singletonList(keyCoder);
+    }
+
+    @Override
+    public List<? extends org.apache.beam.sdk.coders.Coder<?>> getComponents() {
+      return Collections.singletonList(keyCoder);
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      verifyDeterministic(this, "Key coder must be deterministic", keyCoder);
+    }
+
```

Comment:
Please also implement structuralValue, verifyDeterministic, consistentWithEquals, isRegisterByteSizeObserverCheap.

Reply:
Done. Did not use the `StructuralByteArray` for shard id. Seems fine since AutoValue does compare values of bytes array as opposed to object identity.

Reply:
Done.